sycl/include/sycl/ext/oneapi/device_global/device_global.hpp

```diff
@@ -0,0 +1,197 @@
+//==----------- device_global.hpp - SYCL device_global extension -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cstddef>
+#include <type_traits>
+
+#include <sycl/detail/stl_type_traits.hpp>
+#include <sycl/exception.hpp>
+#include <sycl/ext/oneapi/device_global/properties.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+#ifdef __SYCL_DEVICE_ONLY__
+#define __SYCL_HOST_NOT_SUPPORTED(Op)
+#else
+#define __SYCL_HOST_NOT_SUPPORTED(Op)                                          \
+  throw sycl::exception(                                                       \
+      sycl::make_error_code(sycl::errc::feature_not_supported),                \
+      Op " is not supported on host device.");
+#endif
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+namespace detail {
+// Type-trait for checking if a type defines `operator->`.
+template <typename T, typename = void>
+struct HasArrowOperator : std::false_type {};
+template <typename T>
+struct HasArrowOperator<
+    T, sycl::detail::void_t<decltype(std::declval<T>().operator->())>>
+    : std::true_type {};
+
+// Base class for device_global.
+template <typename T, typename PropertyListT, typename = void>
+class device_global_base {
+protected:
+  T *usmptr;
+  T *get_ptr() noexcept { return usmptr; }
+  const T *get_ptr() const noexcept { return usmptr; }
+};
+
+// Specialization of device_global base class for when device_image_scope is in
+// the property list.
+template <typename T, typename... Props>
+class device_global_base<
+    T, properties_t<Props...>,
+    sycl::detail::enable_if_t<properties_t<Props...>::template has_property<
+        device_image_scope_key>()>> {
+protected:
+  T val{};
+  T *get_ptr() noexcept { return &val; }
+  const T *get_ptr() const noexcept { return &val; }
+};
+} // namespace detail
+
+template <typename T, typename PropertyListT = detail::empty_properties_t>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+    // FIXME: Temporary work-around. Remove when fixed.
+    [[__sycl_detail__::global_variable_allowed, __sycl_detail__::device_global]]
+#endif
+    device_global {
+  // This should always fail when instantiating the unspecialized version.
+  static_assert(is_property_list<PropertyListT>::value,
+                "Property list is invalid.");
+};
```

Comment:
I don't think we need extra specialization for this. IMO, just moving the static assert to the class below would work as good.

Reply:
This is the base template and the one below is the specialization. We need to do it this way to unpack the properties inside the property list. However, we still run the risk of a user using the base-template here, e.g. as `device_global<int, double>` which should fail with the "Property list is invalid."

In fact, in all cases where the base template is used, the static_assert will fail, as otherwise it would have gone for the specialization below, but we cannot make it `static_assert(false, ...)` as it will always fail to compile, so we must make it dependent.

Reply:
Can we limit base case to declaration-only instead of definition? If that wouldn't be trivial, I'm fine with the current approach.

Reply:
We could, but I think for the sake of error reporting, having definition with a static assert helps inform the user of what is wrong, for example if they do a `device_global<int, double>`, they'd be told specifically there's a problem with the property list rather than just being told that there is no definition for the class. From their perspective, the `is_property_list<PropertyListT>::value` check is still relevant, even though we know that for this definition it won't ever be `true`.